### PR TITLE
Feat/expose sigs and offset 60 62

### LINKED ISSUE #60: Enable to expose shared signals
Currently, the DSL function expose only accepts forward signals, add the possibility for shared signals.

### PR DESCRIPTION
**Pull request by lancenonce**
Fixed #60 , #61 , and #62 .

Need more clarity on issue 62; how is the offset expressed in the circuit? Is it the `i32` value in `ir::Circuit`'s `pub exposed: Vec<(Column, i32)>` ? Logic should be straight-forward to implement in `expose_first`, `expose_last`, and `expose_step` in `ast::Circuit`.

### CODE REVIEW

File: src/ast.rs
Comment (leolara) [inline src/ast.rs:19]: I don't think we should have a new enum just for the exposed feature, just check in ast::expose could recieve a Queriable, and then panic for when no forward or shared.

Fixed signal cannot be exposed.

File: src/dsl.rs
Comment (leolara) [inline src/dsl.rs:59]: @lancenonce I think it is better to pass Queriable to ast::expose, so we don't need to add an enum just for the expose feature

File: src/ast.rs
Comment (leolara) [inline src/ast.rs:198]: In the AST you just need to store the information, then the compiler will convert it to column, rotation

Comment (leolara) [reply src/ast.rs:198]: So you can store with the  ExposeOffset enum. And perhaps you don't need three methods then.

**Comment by leolara** [comment]
@lancenonce there is no logic in the ast, it is just a container with information. You should just store the signal, offset to be exposed. Then the compiler should convert that in the IR to (Column,rotation).

The backend is already taking the (Column,rotation) so it should work from there without modification.

**Comment by lancenonce** [comment]
Thank you for these helpful comments @leolara ! I'm on it.

**Comment by lancenonce** [comment]
Updates:

- We no longer use a `Signal` enum, just a Queriable that checks if the exposed signals are Forward / Shared
- We compile the exposed signals by returning the column and rotation depending on the signal and the offset passed from `ast`
- We store exposed signals in ast::Circuit with a `Vec<(Queriable<F>, ExposeOffset)>`

File: src/dsl.rs
Comment (leolara) [inline src/dsl.rs:166]: I think this enum should be in the ast.rs module

File: src/compiler.rs
Comment (leolara) [inline src/compiler.rs:334]: So if First should be equivalent to step=0 and Last to step=num_steps - 1

**Review by leolara — CHANGES_REQUESTED** [review]

File: src/compiler.rs
Comment (leolara) [inline src/compiler.rs:347]: @lancenonce the shared signal is still with the wrong values for rotation

Comment (lancenonce) [reply src/compiler.rs:347]: Fixed, thanks @leolara !

File: src/frontend/pychiquito.rs
Comment (lancenonce) [inline src/frontend/pychiquito.rs:71]: @qwang98 this is where the breaking change lies. Queriables do not implement any serde traits.

**Review by leolara — APPROVED** [review]

File: src/frontend/pychiquito.rs
Comment (qwang98) [reply src/frontend/pychiquito.rs:71]: I merged over this branch and I'm working on this :)

File: src/ast.rs
Comment (qwang98) [inline src/ast.rs:122]: Checking for Forward and Shared should already be done in the dsl `expose` function, so here we can simply push the `(signal, offset)`.

File: src/dsl.rs
Comment (qwang98) [inline src/dsl.rs:62]: Here in `expose` we should check for Forward and Share as you've done, and then call `expose` in the AST rather than directly pushing to `exposed` of AST.

In general, we want functions that modifies AST to remain in AST, so that DSL can do simple calls of AST functions. Let me know if this makes sense!

**Review by leolara — DISMISSED** [review]
Waiting for Steve comments to be done

File: src/dsl.rs
Comment (leolara) [reply src/dsl.rs:62]: @qwang98 good catch

File: src/ast.rs
Comment (leolara) [reply src/ast.rs:122]: I think it is better to do it in AST, if we are going to do it only in one place. As if someone modifies the AST without the DSL the check should be done

**Comment by lancenonce** [comment]
Changes should be good to go, please let me know what you think of the new Error enum, I'm thinking we can use this for more precise error handling @qwang98 @leolara

**Comment by lancenonce** [comment]
@qwang98 to merge pychiquito fixes after this is merged in.

**Comment by qwang98** [comment]
> Changes should be good to go, please let me know what you think of the new Error enum, I'm thinking we can use this for more precise error handling @qwang98 @leolara

New fixes look good to me!

Re error handling, we've been panic'ing everything so far, which is not the proper way lol. @leolara I think we'd eventually need precise error handling everywhere. What do you think?

**Comment by qwang98** [comment]
I fixed frontend related stuffs and should be ready for merge once @leolara approves again.

**Comment by leolara** [comment]
> > Changes should be good to go, please let me know what you think of the new Error enum, I'm thinking we can use this for more precise error handling @qwang98 @leolara
> 
> New fixes look good to me!
> 
> Re error handling, we've been panic'ing everything so far, which is not the proper way lol. @leolara I think we'd eventually need precise error handling everywhere. What do you think?

Panicking is good if the user is creating a chiquito program that is not compilable

**Comment by leolara** [comment]
> I fixed frontend related stuffs and should be ready for merge once @leolara approves again.

@qwang98 you are now the the reviewer, it can be merged when you approve it

**Comment by qwang98** [comment]
> > > Changes should be good to go, please let me know what you think of the new Error enum, I'm thinking we can use this for more precise error handling @qwang98 @leolara
> > 
> > 
> > New fixes look good to me!
> > Re error handling, we've been panic'ing everything so far, which is not the proper way lol. @leolara I think we'd eventually need precise error handling everywhere. What do you think?
> 
> Panicking is good if the user is creating a chiquito program that is not compilable

kk I reverted back to prior panic message just to be consistent throughout the program. We can think about error handling in the future. I think a systematic approach should still be good.

**Review by qwang98 — APPROVED** [review]

**Pull request merged**
